main/src/library/Traversable.flix

```diff
@@ -0,0 +1,170 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+///
+/// A type class for data structures that can be traversed with an applicative functor.
+///
+pub class Traversable[t : Type -> Type] with Functor[t], Foldable[t] {
+
+    ///
+    /// Returns the result of applying the applicative mapping function `f` to all the elements of the
+    /// data structure `t`.
+    ///
+    pub def traverse(f: a -> m[b] & ef, t: t[a]): m[t[b]] & ef with Applicative[m]
+
+    ///
+    /// Returns the result of running all the actions in the data structure `t`.
+    ///
+    pub def sequence(t: t[m[a]]): m[t[a]] with Applicative[m] =
+        Traversable.traverse(identity, t)
+
+    ///
+    /// Traversing with the identity function wrapped into an applicative preserves the container `t`.
+    ///
+    law identity: forall(t: t[a], f: a -> m[a]) with Eq[m[t[a]]], Applicative[m] . Traversable.traverse(f, t) == Applicative.point(t)
+
+    ///
+    /// sequence identity.
+    ///
+    law identity: forall(t: t[a], f: a -> m[a]) with Eq[m[t[a]]], Applicative[m] . Traversable.sequence(Functor.map(f, t)) == Applicative.point(t)
+
+}
+
+
+namespace Traversable {
+
+    ///
+    /// Returns the result of applying the applicative mapping function `f` to all the elements of the
+    /// data structure `t`.
+    ///
+    /// `for` is `traverse` with it's arguments flipped.
+    ///
+    pub def for(t: t[a], f: a -> m[b] & ef): m[t[b]] & ef with Applicative[m], Traversable[t] =
+        Traversable.traverse(f, t)
+
+    ///
+    /// Returns the result of applying `f` to the traversable structure `t` and the initial state `acc`.
+    /// The result is a pair of the final state and the updated copy of the structure.
+    ///
+    /// `mapAccumLeft` is essentially the combination of `map` and `foldLeft` - like map it returns an updated copy
+    /// of the intial structure, like `foldLeft` it passes an updating accumulator through each step of the traversal.
+    ///
+    pub def mapAccumLeft(f: acc -> a -> (acc, b) & ef, start: acc, t: t[a]): (acc, t[b]) & ef with Traversable[t] =
+        runContStateLeft(traverse(x1 -> ContStateLeft((k, s) -> let (s1, b) = f(s, x1); k(b, s1)), t), start) as & ef
+
+    ///
+    /// Helper type for `mapAccumLeft`.
+    ///
+    /// ContStateLeft the type of the State monad implemented in CPS. In this case we just need
+    /// `Applicative.ap` so no Monad instance is defined.
+    ///
+    /// A simple implementation of the State monad bursts the stack when `mapAccumLeft` is run on large
+    /// collections, so use ContState the CPS-State monad instead.
+    ///
+    /// We use an Impure continuation so we can internally cast the `ef` of `mapAccumLeft` and not lose
+    /// impurity information.
+    ///
+    opaque type ContStateLeft[ka, s, a] = (a -> s -> ka & Impure) -> s -> ka & Impure
```

Comment:
What's the reason that the function must be `& Impure`. 

Usually functions are either pure or effect polymorphic. I suspect that ContStateLeft should have an extra `ef` parameter. Would that work?

Reply:
I think you should be able to do that now, but you have to annotate the type parameters, e.g. `enum Foo[a: Type, ef: Bool]`.

Reply:
I'm still struggling to write monads with an `ef` parameter - see long note below.